### PR TITLE
Switch to `streamx` and port tests to Brittle

### LINKED ISSUE #38: Move to streamx?
Would it be reasonable to move utp-native from readable-stream to streamx?

### PR DESCRIPTION
**Pull request by kasperisager**
Closes #38
Closes #42

### CODE REVIEW

File: test/net.js [hunk 1]
```diff
@@ -327,11 +330,12 @@ tape('emits close', function (t) {
     server.close()
     t.ok(serverClosed)
     t.ok(clientClosed)
-    t.end()
   }
 })
 
-tape('flushes', function (t) {
+test('flushes', function (t) {
```
Comment (kasperisager) [inline test/net.js]: This test currently causes a segfault:

```
PID 11349 received SIGSEGV for address: 0x0
0   segfault-handler.node               0x0000000112e5f09a _ZL16segfault_handleriP9__siginfoPv + 298
1   libsystem_platform.dylib            0x00007fff204ecd7d _sigtramp + 29
2   ???                                 0x0000000000000026 0x0 + 38
3   node                                0x000000010e6da35b _ZN4node6Buffer4DataEN2v85LocalINS1_5ValueEEE + 23
4   node                                0x000000010e6d2388 napi_get_buffer_info + 51
5   utp_native.node                     0x0000000112e3889a _ZL13on_utp_acceptP22utp_callback_arguments + 346
6   utp_native.node                     0x0000000112e40e7d _Z18utp_call_on_acceptP18struct_utp_contextP9UTPSocketPK8sockaddrj + 45
7   utp_native.node                     0x0000000112e3f33d utp_process_udp + 1405
8   utp_native.node                     0x0000000112e390b0 _ZL10on_uv_readP8uv_udp_slPK8uv_buf_tPK8sockaddrj + 144
9   libuv.1.dylib                       0x0000000110a1e375 uv__udp_io + 350
10  libuv.1.dylib                       0x0000000110a21773 uv__io_poll + 1861
11  libuv.1.dylib                       0x0000000110a11e46 uv_run + 359
12  node                                0x000000010e6470d7 _ZN4node13SpinEventLoopEPNS_11EnvironmentE + 255
13  node                                0x000000010e7333ef _ZN4node16NodeMainInstance3RunEPiPNS_11EnvironmentE + 97
14  node                                0x000000010e733082 _ZN4node16NodeMainInstance3RunEPKNS_16EnvSerializeInfoE + 126
15  node                                0x000000010e6d0039 _ZN4node5StartEiPPc + 216
16  libdyld.dylib                       0x00007fff204c2f3d start + 1
17  ???                                 0x0000000000000002 0x0 + 2
zsh: segmentation fault  node test/net.js
```

File: test/net.js [hunk 2]
```diff
@@ -357,7 +360,9 @@ tape('flushes', function (t) {
   })
 })
 
-tape('close waits for connections to close', function (t) {
+test('close waits for connections to close', function (t) {
```
Comment (kasperisager) [inline test/net.js]: This test currently causes a segfault:

```
PID 11370 received SIGSEGV for address: 0x0
0   segfault-handler.node               0x000000010c2d509a _ZL16segfault_handleriP9__siginfoPv + 298
1   libsystem_platform.dylib            0x00007fff204ecd7d _sigtramp + 29
2   ???                                 0x0000000000000026 0x0 + 38
3   node                                0x0000000107b5235b _ZN4node6Buffer4DataEN2v85LocalINS1_5ValueEEE + 23
4   node                                0x0000000107b4a388 napi_get_buffer_info + 51
5   utp_native.node                     0x000000010c2ae89a _ZL13on_utp_acceptP22utp_callback_arguments + 346
6   utp_native.node                     0x000000010c2b6e7d _Z18utp_call_on_acceptP18struct_utp_contextP9UTPSocketPK8sockaddrj + 45
7   utp_native.node                     0x000000010c2b533d utp_process_udp + 1405
8   utp_native.node                     0x000000010c2af0b0 _ZL10on_uv_readP8uv_udp_slPK8uv_buf_tPK8sockaddrj + 144
9   libuv.1.dylib                       0x0000000109e99375 uv__udp_io + 350
10  libuv.1.dylib                       0x0000000109e9c773 uv__io_poll + 1861
11  libuv.1.dylib                       0x0000000109e8ce46 uv_run + 359
12  node                                0x0000000107abf0d7 _ZN4node13SpinEventLoopEPNS_11EnvironmentE + 255
13  node                                0x0000000107bab3ef _ZN4node16NodeMainInstance3RunEPiPNS_11EnvironmentE + 97
14  node                                0x0000000107bab082 _ZN4node16NodeMainInstance3RunEPKNS_16EnvSerializeInfoE + 126
15  node                                0x0000000107b48039 _ZN4node5StartEiPPc + 216
16  libdyld.dylib                       0x00007fff204c2f3d start + 1
17  ???                                 0x0000000000000002 0x0 + 2
zsh: segmentation fault  node test/net.js
```

File: index.js [hunk 3]
```diff
@@ -1,265 +1,270 @@
...
 
-UTP.prototype._onclose = function () {
-  binding.utp_napi_destroy(this._handle, this._sent.map(toHandle))
-  this._handle = null
-  this.emit('close')
+module.exports = {
+  Socket,
+  createServer,
+  connect
 }
```
Comment (mafintosh) [inline index.js]: module.exports = Socket and then we add the other methods as static exports?

File: lib/connection.js [hunk 4]
```diff
@@ -13,220 +12,218 @@ const UTP_ERRORS = [
...
-  this._buffer = Buffer.allocUnsafe(this._buffer.length)
-  this._offset = 0
-  return this._buffer
-}
+    if (bufs.length > 256) return this._write(Buffer.concat(bufs), cb)
 
-Connection.prototype._onread = function (size) {
-  if (!this._connected) this._onconnect() // makes the server wait for reads before writes
-  if (this._timeout) this._timeout.refresh()
+    if (!binding.utp_napi_connection_writev(this._handle, bufs)) {
```
Comment (mafintosh) [inline lib/connection.js]: this needs the !this._connected check also cause streamx will eagerly use this hook

File: lib/connection.js [hunk 5]
```diff
@@ -13,220 +12,218 @@ const UTP_ERRORS = [
...
 
-  const bufs = new Array(datas.length)
-  for (var i = 0; i < datas.length; i++) bufs[i] = datas[i].chunk
+  _read (cb) {
+    // TODO: backpressure
+    // cb(null)
+  }
 
-  if (bufs.length > 256) return this._write(Buffer.concat(bufs), null, cb)
+  _write (data, cb) {
```
Comment (mafintosh) [inline lib/connection.js]: in streamx, having writev only is enough so this should be removed

File: lib/connection.js [hunk 6]
```diff
@@ -13,220 +12,218 @@ const UTP_ERRORS = [
...
+  address () {
+    if (this.destroyed) return null
+    return this._utp.address()
+  }
 
-  const bufs = new Array(datas.length)
-  for (var i = 0; i < datas.length; i++) bufs[i] = datas[i].chunk
+  _read (cb) {
+    // TODO: backpressure
+    // cb(null)
```
Comment (mafintosh) [inline lib/connection.js]: you have to call the cb here always

File: test/net.js [hunk 7]
```diff
@@ -432,16 +430,28 @@ tape('timeout', function (t) {
...
+
+test('exception in connection listener', async function (t) {
+  t.plan(1)
+
+  const server = utp.createServer(function (socket) {
+    socket.destroy()
+    throw new Error('disconnect')
+  })
+
+  process.once('uncaughtException', () => {
```
Comment (kasperisager) [inline test/net.js]: @davidmarkclements Brittle sets up an uncaught exception capture callback so this event isn't emitted. Would it be possible to have Brittle emit it and bail out if it was handled by a listener?

File: test/net.js (hunk 1 above)
Comment (kasperisager) [reply test/net.js]: @mafintosh This test now fails as only the first write makes it through. Is this the intended behaviour of streamx?

File: test/net.js [hunk 8]
```diff
@@ -1,447 +1,533 @@
...
-  server.listen(0, function () {
-    var socket = utp.connect(server.address().port)
+  await close
+}))
 
-    socket.write('hi')
-    socket.once('connect', function () {
-      socket.end()
-    })
+test.skip('client destroys on connect', (t) => withServer(t, async (server) => {
```
Comment (davidmarkclements) [inline test/net.js:102]: can we add a comment for why this is skipped

File: test/net.js (hunk 7 above)
Comment (davidmarkclements) [reply test/net.js]: yes absolutely, would you be okay to make that PR to brittle?

File: test/sockets.js [hunk 9]
```diff
@@ -1,130 +1,129 @@
...
-      t.end()
+      t.alike(data, Buffer.from('hi'))
     })
   })
 })
 
-tape('both ends write first', function (t) {
-  var missing = 2
-  const socket = utp()
+test.skip('both ends write first', async (t) => {
```
Comment (davidmarkclements) [inline test/sockets.js]: can we comment on this skip as well

Comment (kasperisager) [reply test/sockets.js]: That one was unintentional, thanks for the heads up!

**Comment by davidmarkclements** [comment]
@kasper this is great work, haven't approved because I assume we need that change to brittle first

File: test/net.js (hunk 7 above)
Comment (kasperisager) [reply test/net.js]: https://github.com/davidmarkclements/brittle/pull/12

**Comment by kasperisager** [comment]
@davidmarkclements This should be good to go now that Brittle has been updated.

**Comment by kasperisager** [comment]
@mafintosh Should we move ahead with this or close it?

**Comment by mafintosh** [comment]
We moved to udx

**Pull request closed without merging**